### DISCUSSION TITLE
Async Tests for Websocket not possible with HTTPX

### DISCUSSION THREAD

**Discussion by aayushchhabra1999** [fastapi/fastapi#12242] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
"""Testing client for the FastAPI application."""

from collections.abc import AsyncGenerator

import pytest_asyncio
from asgi_lifespan import LifespanManager
from fastapi import FastAPI
from httpx import ASGITransport, AsyncClient

from app_factory.factory import create_fast_app
from dependency_injection.container import container, test_container
from logger import get_logger

logger = get_logger(__name__)


@pytest_asyncio.fixture(scope="module")
async def kniru_api_app() -> AsyncGenerator[FastAPI, None]:
    """Create a FastAPI application for the Kniru API."""
    kniru_api_app = create_fast_app()
    kniru_api_app.dependency_overrides[container] = test_container
    async with LifespanManager(kniru_api_app) as manager:
        yield manager.app


@pytest_asyncio.fixture(scope="module")
async def kniru_api_async_client(kniru_api_app: FastAPI) -> AsyncGenerator[AsyncClient, None]:
    """Create a test client for the Kniru API."""
    async with AsyncClient(
        transport=ASGITransport(app=kniru_api_app),
        base_url="http://test",
    ) as async_client:
        yield async_client

---

@pytest.mark.asyncio
@pytest.mark.usefixtures("create_current_user")
async def test_chat_websocket_connection(
    kniru_api_async_client: AsyncClient,
    bearer_auth_header: dict,
):
    """Test the WebSocket connection for chat."""
    async with kniru_api_async_client.websocket_connect(
        "/chat/ws",
        headers=bearer_auth_header,
    ) as websocket:
        # I will break because httpx doesn't support websockets
        # 'AsyncClient' object has no attribute 'websocket_connect'
```

### Description

The docs suggest using httpx -> AsyncClient for doing async tests. However, httpx doesn't support websockets. What is the correct way to do websocket testing.

Our requirement: We want to create tests where we can test websocket flows which require establishing a websocket connection, then sending a json body on the websocket, and then inspecting the database to make sure the proper set of changes took place.

### Operating System

macOS

### FastAPI Version

0.115.0

### Pydantic Version

2.9.1

### Python Version

Python 3.11.9

**Comment by Kludex** [comment, marked as answer]
Use https://github.com/frankie567/httpx-ws.
